Adapt processors and pid_max to macOS

simonnagl opened the pull request:
@vlvkobal could you please review that? Thank you for starting porting netdata to macOS. This is fantastic!

I moved `getsysctl()` to commons.c because otherwise I get this error:

```
clang: warning: argument unused during compilation: '-pthread'
Undefined symbols for architecture x86_64:
  "_getsysctl", referenced from:
      _get_system_cpus in common.o
      _get_system_pid_max in common.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [apps.plugin] Error 1
make[1]: *** [all-recursive] Error 1
```

Can somebody help me with that?

vlvkobal commented:
I think it's better to leave `getsysctl()` in `macos_sysctl.c` and add `extern` qualifier to it's definition in `plugin_macos.h`

vlvkobal commented on src/common.c:
```diff
@@ -1129,10 +1141,24 @@ long get_system_cpus(void) {
+    #ifdef __APPLE__
+        int32_t tmp_pid_max;
+
+        if (unlikely(GETSYSCTL("kern.maxproc", tmp_pid_max))) {
```
`kern.maxproc` is not the maximum pid but maximum number of processes. On my macOS VPS kern.maxproc = 532. In FreeBSD we have `kern.pid_max`. I can't find nothing similar in macOS sysctls.

simonnagl replied:
You are absolutely right. This was my fault. I found this:
http://apple.stackexchange.com/questions/51119/whats-the-maximum-pid-for-mac-os-x

On my device I found no PID higher than 99999. The current default is 32768.

Currently the highest PID on my running macOS is 95999. Maybe we should set 99999 as the default for macOS.
@ktsaou Should I set it also as default if not found on another os?

vlvkobal commented on src/common.c:
```diff
@@ -1100,6 +1100,18 @@ int processors = 1;
+    #ifdef __APPLE__
+        int32_t tmp_processors;
+
+        if (unlikely(GETSYSCTL("hw.logicalcpu", tmp_processors))) {
```
Have you considered `hw.ncpu`? Is `hw.logicalcpu` better in this case?

simonnagl replied:
From man 3 sysctl:

> hw.ncpu
             The number of cpus. This attribute is deprecated and it is recom-
             mended that hw.logicalcpu, hw.logicalcpu_max, hw.physicalcpu, or
             hw.physicalcpu_max be used instead.

vlvkobal replied:
You are right. Thank you for the reference.

simonnagl replied:
So `hw.ncpu` is not what we want to use. I think `hw.logicalcpu ` wins in favor of `hw.physicalcpu ` because that reflects whats in `proc/stat`
I am not sture if `hw.logicalcpu_max` is what we want here because `hw.logicalcpu` may vary depending on the current power management mode. What do you think?

vlvkobal replied:
I think it is OK to use `hw.logicalcpu`

simonnagl commented:
> I think it's better to leave getsysctl() in macos_sysctl.c and add extern qualifier to it's definition in plugin_macos.h.

This was my initial intention but the build fails. I'll push the failing state so you can see it.

simonnagl commented:
6fc09b9 fails on my Mac. @vlvkobal Can you verify this?

vlvkobal commented:
Yes. It fails. I'll look at it a bit later.

vlvkobal commented:
Switch off `apps.plugin` compiling for macOS in `src/Makefile.am`. It isn't working anyway. We will have to rewrite it using native API for both macOS and FreeBSD.
```
if MACOS
else
plugins_PROGRAMS = apps.plugin
endif
```

simonnagl commented:
@vlvkobal Thank you for helping me out!

When you say it it seems obvious. I pushed the changes. It now builds and can be merged. Can you verify that?

vlvkobal commented:
It works.

It's better to include a comment about where did you get the `pid_max` number from right into the `common.c`. Instead of URL you can use something more general like "bsd/sys/proc_internal.h in XNU sources".

Try to avoid `No newline at end of file` warnings (http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file).

ktsaou commented:
guys when you want something merged, let me know.

simonnagl commented:
@vlvkobal I added the comment rebased and fixed net end of file issues.

If you think this is ready tell @ktsaou. I think the macOS port is your baby. Thanks for that!

vlvkobal commented:
I don't have any other suggestions. I think it can be merged.

ktsaou commented:
merged!
